<a href="https://colab.research.google.com/github/amilsiddik/Deep-Learning/blob/main/02_Multilayer_Perceptron.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

>Pada Part 1 kita sudah mengenal apa itu neural network, activation function dan sudah mencoba implementasi forward propagation untuk melakukan regresi terhadap fungsi linear f(x) = 3x + 2

>Fungsi linear diatas adalah fungsi yang sangat simple sehingga dengan menggunakan 2 layer (Input dan Output) saja kita sudah bisa menyelesaikan permasalahan tersebut. Lalu bagaimana dengan fungsi non-linear? Tentu saja kita tidak bisa menggunakan arsitektur 2 layer tersebut.

>Sehingga untuk non-linear regression kita membutuhkan setidaknya 3 layer neural network atau yang biasa disebut Multilayer Perceptron (MLP) atau Fully-Connected Layer dengan menggunakan non-linear activation function pada seluruh neuron di hidden layer.

Kita akan mencoba melakukan forward pass pada MLP masih dengan Numpy saja. Untuk contoh kasusnya adalah kita akan melakukan regresi untuk data yang sebenarnya adalah sebuah fungsi non-linear sebagai berikut:
> f(x)=\sqrt(2x^2 +1)

Sedangkan arsitektur neural networknya terdiri dari :
> 1 node pada input layer

> 8 node pada hidden layer pertama (ReLU)

> 1 node pada output layer (Linear)

**Forward Propagation**

In [1]:
def forwardPass(inputs, weight, bias, activation = 'linear'):
	w_sum = np.dot(inputs, weight) + bias

	if activation is 'relu' :
		# ReLU Activation f(x) = max(0, x)
		act = np.maximum(w_sum, 0)
	else :
		# Linear Activation f(x) = x
		act = w_sum

	return act

In [2]:
import numpy as np

def forwardPass(inputs, weight, bias, activation = 'linear'):
	w_sum = np.dot(inputs, weight) + bias

	if activation is 'relu' :
		# ReLU Activation f(x) = max(0, x)
		act = np.maximum(w_sum, 0)
	else :
		# Linear Activation f(x) = x
		act = w_sum

	return act

# Pre-Trained Weights & Biases after Training
W_H = np.array([[0.00192761, -0.78845304, 0.30310717, 0.44131625, 0.32792646, -0.02451803, 1.43445349, -1.12972116]])
b_H = np.array([-0.02657719, -1.15885878, -0.79183501, -0.33550513, -0.23438406, -0.25078532, 0.22305705, 0.80253315])

W_o = np.array([[-0.77540326], [ 0.5030424 ], [ 0.37374797], [-0.20287184], [-0.35956827], [-0.54576212], [ 1.04326093], [ 0.8857621 ]])
b_o = np.array([ 0.04351173])

# Initialize Input Data
inputs = np.array([[-2], [0], [2]])

#Output of Hidden Layer
h_out = forwardPass(inputs, W_H, b_H, 'relu')

print('Hidden Layer Output (ReLU)')
print('================================')
print(h_out, "\n")

# Output of Output Layer
o_out = forwardPass(h_out, W_o, b_o, 'linear')

print('Output Layer Output (Linear)')
print('================================')
print(o_out, "\n")


Hidden Layer Output (ReLU)
[[0.         0.4180473  0.         0.         0.         0.
  0.         3.06197547]
 [0.         0.         0.         0.         0.         0.
  0.22305705 0.80253315]
 [0.         0.         0.         0.54712737 0.42146886 0.
  3.09196403 0.        ]] 

Output Layer Output (Linear)
[[2.96598907]
 [0.98707188]
 [3.00669343]] 



Pada percobaan non-linear regression kali ini kita akan melakukan perdiksi nilai dari -2, 0 dan 2. Output yang dihasilkan seharusnya adalah 3, 1, 3 dan hasil prediksi adalah 2.96598907, 0.98707188 dan 3.00669343. Masih ada sedikit error tapi paling tidak hasil diatas menunjukkan bahwa MLP dapat melakukan regresi terhadap fungsi non-linear dengan cukup baik.